src/compiler/scala/tools/nsc/transform/Delambdafy.scala

```diff
@@ -46,9 +48,9 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
 
     val typer = localTyper
 
-    // we need to know which methods refer to the 'this' reference so that we can determine
-    // which lambdas need access to it
-    val thisReferringMethods: Set[Symbol] = {
+    // we need to know which methods refer to the 'this' reference so that we can determine which lambdas need access to it
+    // TODO: this looks expensive, so I made it a lazy val. Can we make it more pay-as-you-go / optimize for common shapes?
```

Comment:
I think we could reduce prune the traversal with https://github.com/retronym/scala/commit/571da7d59522775747a5b2b67031677952638092

Under that change, we still create static lambda impl methods for:

```
⚡ cat sandbox/test.scala && qscalac sandbox/test.scala && javap C
class C {
  def foo = {
    () => () => () => 42
  }
}
Compiled from "test.scala"
public class C {
  public scala.Function0<scala.Function0<scala.Function0<java.lang.Object>>> foo();
  public static final int C$$$anonfun$3();
  public static final scala.Function0 C$$$anonfun$2();
  public static final scala.Function0 C$$$anonfun$1();
  public C();
}
```